packages/protocol/test/unit/utils/StreamPartitionID.test.ts

```diff
@@ -0,0 +1,188 @@
+    const PARTITION = 1
+
+    it('can use object destructuring', () => {
+        const spid = new SPID(STREAM_ID, PARTITION)
+        const { id, partition, key } = spid
+        expect(id).toBe(STREAM_ID)
+        expect(partition).toBe(PARTITION)
+        expect(key).toBe(spid.toString())
+    })
+
+    it('defaults partition to 0', () => {
+        const spid = new SPID(STREAM_ID)
+        expect(spid.id).toEqual(STREAM_ID)
+        expect(spid.partition).toBe(0)
+    })
+
+    it('lowercases streamId', () => {
+        const spid = new SPID(STREAM_ID.toUpperCase())
+        expect(spid.id).toEqual(STREAM_ID)
+        expect(spid.partition).toBe(0)
+    })
+
+    it('can use array destructuring with split', () => {
+        const spid = new SPID(STREAM_ID, PARTITION)
+        const [ id, partition ] = spid.split()
+        expect(id).toBe(STREAM_ID)
+        expect(partition).toBe(PARTITION)
+    })
+
+    it('can get key from string', () => {
+        const spid = new SPID(STREAM_ID, PARTITION)
+        const key = SPID.toKey(spid.toString())
+        expect(key).toBe(spid.key)
+    })
+
+    it('requires valid streamId', () => {
+        const badTypes = [
+            undefined,
+            null,
+            {},
+            /regex/,
+            [],
+            [1,2,3],
+            Number.NaN,
+            Symbol('bad streamid'),
+            BigInt(1),
+            BigInt(0),
+            function() {},
+            new Uint8Array()
+        ]
+        const badStreamIds = [
+            '', // too short
+            0, // not a string
+        ]
+        for (const streamId of [...badTypes, ...badStreamIds]) {
+            expect(() => {
+                // @ts-expect-error testing bad input
+                new SPID(streamId, PARTITION)
+            }).toThrow()
+        }
+    })
+
+    it('requires valid partition', () => {
+        const badTypes = [
+            null,
+            {},
+            '',
+            /regex/,
+            [],
+            [1,2,3],
+            Symbol('bad streamid'),
+            BigInt(1),
+            BigInt(0),
+            function() {},
+            new Uint8Array()
+        ]
+
+        const badPartitions = [
+            '0', // not a number
+            -1, // not positive
+            0.1, // not integer
+            -0.1, // negative and not integer
+            Number.POSITIVE_INFINITY, // Too big
+            Number.NEGATIVE_INFINITY, // Not positive
+            Number.MAX_SAFE_INTEGER + 1, // Not safe
+            Number.NaN, // NaN
+        ]
+
+        for (const partition of [...badTypes, ...badPartitions]) {
+            expect(() => {
+                // @ts-expect-error testing bad input
+                new SPID(STREAM_ID, partition)
+            }).toThrow()
+        }
+    })
+
+    describe('from', () => {
+        it('returns self if already a spid', () => {
+            const spid = new SPID(STREAM_ID, PARTITION)
+            expect(SPID.from(spid)).toBe(spid)
```

Comment:
I'm assuming the `toBe` here is checking equality of identity as opposed to value equality, right?